### PR TITLE
Feature withDirty HOC

### PR DESCRIPTION
**Pull request by GermainBergeron**
Add new HOC for withDirty
Modify withEditing HOC to use withDirty which is more modular
Add new Prevent Navigate HOC

### CODE REVIEW

File: src/components/modal/ModalComposite.tsx [hunk 1]
```diff
@@ -99,6 +106,16 @@ export class ModalComposite extends React.PureComponent<IModalCompositeProps & P
         window.clearTimeout(this.timeoutId);
     }
 
+    private onRequestClose() {
+        if (this.props.validateShouldNavigate) {
+            if (this.props.validateShouldNavigate(this.props.isDirty)) {
+                this.props.onClose();
+            }
+        } else {
+            callIfDefined(this.props.onClose);
```
Comment (gdostie) [inline src/components/modal/ModalComposite.tsx]: You check if onClose is defined here, but not in the previous if, is this expected?

File: src/components/modal/ModalComposite.tsx [hunk 2]
```diff
@@ -23,6 +24,7 @@ export interface IModalCompositeOwnProps extends IModalOwnProps, IModalHeaderOwn
     modalFooterChildren?: React.ReactNode;
     modalFooterClasses?: IClassName;
     isPrompt?: boolean;
+    validateShouldNavigate?: (isDirty: boolean) => void;
```
Comment (gdostie) [inline src/components/modal/ModalComposite.tsx]: `validateShouldNavigate` must return a boolean
```suggestion
    validateShouldNavigate?: (isDirty: boolean) => boolean;
```

File: src/components/modal/examples/ModalCompositeConnectedExamples.tsx [hunk 3]
```diff
@@ -13,14 +19,45 @@ const modalId: string = 'modal-composite-connected';
...
-
 const mapDispatchToProps = (dispatch: (action: IReduxAction<IModalActionPayload>) => void): IModalExamplesProps => ({
     openModal: (id: string) => dispatch(openModal(id)),
     closeModal: (id: string) => dispatch(closeModal(id)),
 });
 
-@ReduxConnect(mapStateToProps, mapDispatchToProps)
+const lorem = loremIpsum({count: 10});
+
+@ReduxConnect(undefined, mapDispatchToProps)
```
Comment (gdostie) [inline src/components/modal/examples/ModalCompositeConnectedExamples.tsx]: You could define the mapDispatchToProps as an object instead (see [react-redux's doc](https://react-redux.js.org/using-react-redux/connect-mapdispatch#defining-mapdispatchtoprops-as-an-object) about it, it is actually the recommended way to use mapDispatchToProps).
```suggestion
@ReduxConnect(null, {openModal, closeModal})
```
And you can remove the `mapDispatchToProps` function declaration.

File: src/components/modal/examples/ModalCompositeConnectedExamples.tsx [hunk 4]
```diff
@@ -116,6 +153,14 @@ export class ModalCompositeConnectedExamples extends React.Component<IModalExamp
...
                         />
                     </div>
+
+                    <div className='mt3'>
+                        <label className='form-control-label'>Modal With a Prevent Navigation</label>
+                        <div className='form-group'>
+                            <button className='btn' onClick={() => this.props.openModal('modal-composite-2')}>Open Modal</button>
+                            <ComponentWithPreventNavigationHOC />
+                        </div>
+                    </div>
```
Comment (gdostie) [inline src/components/modal/examples/ModalCompositeConnectedExamples.tsx]: I don't see this example in the demo, maybe it has not finished building yet?

File: src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    stay: 'Cancel',
+};
+
+export const modalWithPreventNavigation = <T, R = any>(config: IWithPreventNavigationConfig) => (Component: React.ComponentClass<T, R>): React.ComponentClass<T & Partial<IWithPreventNavigationInjectedProps>, R> => {
+    const mapDispatchToProps = (dispatch: (action: IReduxAction<IModalActionPayload>) => void): IWithPreventNavigationDispatchProps => ({
+        closeModal: (id: string) => dispatch(closeModal(id)),
+    });
+
+    @ReduxConnect(undefined, mapDispatchToProps)
+    class ModalWithPreventNavigation extends React.PureComponent<IWithPreventNavigationDispatchProps, IWithPreventNavigationState> {
```
Comment (gdostie) [inline src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx]: You could put the initial state here
```tsx
state = {
    showPrevent: false,
}
```
That way you could remove the constructor.

File: src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    return isDirty && (
+                        <PreventNavigationPrompt
+                            id={`prevent-navigation-${config.id}`}
+                            isOpen={this.state.showPrevent}
+                            title={title}
+                            onStay={() => this.setState({showPrevent: false})}
+                            onClose={() => {
+                                this.setState({showPrevent: false});
+                                this.props.closeModal(config.id);
+                            }
```
Comment (gdostie) [inline src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx]:
```suggestion
                            }}
```

File: src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                            }
```
Comment (gdostie) [inline src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx]:
```suggestion

```

File: src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    exit: 'Exit without applying changes',
+    stay: 'Cancel',
+};
+
+export const modalWithPreventNavigation = <T, R = any>(config: IWithPreventNavigationConfig) => (Component: React.ComponentClass<T, R>): React.ComponentClass<T & Partial<IWithPreventNavigationInjectedProps>, R> => {
+    const mapDispatchToProps = (dispatch: (action: IReduxAction<IModalActionPayload>) => void): IWithPreventNavigationDispatchProps => ({
+        closeModal: (id: string) => dispatch(closeModal(id)),
+    });
+
+    @ReduxConnect(undefined, mapDispatchToProps)
```
Comment (gdostie) [inline src/hoc/withPreventNavigation/modalWithPreventNavigation.tsx]: Same here, you could define mapDispatchToProps as an object instead
```suggestion
    @ReduxConnect(null, {closeModal})
```

File: src/components/modal/examples/ModalCompositeConnectedExamples.tsx (hunk 4 above)
Comment (GermainBergeron) [reply src/components/modal/examples/ModalCompositeConnectedExamples.tsx]: It looks like an issue with the build 🤔

**Pull request merged**